멀티 모듈 프로젝트 구성 및 인증 기능 구현

i960107 opened the pull request:
## 모듈 구분
1. 모듈은 두개의 레이어로 구분한다. 실행 app 모듈, 메인 모듈로 구분한다.
2. 실행 app 모듈은 실행 가능한 jar파일로 빌드되며 app을 붙여 구분하였다. app이 붙지 않은 모듈은 자체 실행 불가능하다.
4. 모듈은 필요한 부분만 조합해서 사용할 수 있도록 관심사에 따라 분리한다.
  - **app-api** : api 어플리케이션 실행 모듈. domain 모듈을 사용. controller, service 레이어 포함
  - **app-batch**: batch 실행 모듈. domain 모듈을 사용.
  - **domain**: entity, repository 포함. db와의 연결을 담당.
 4. app-api와 app-batch에 공통적으로 적용되어야 하는 에러코드나 인증로직은 많지 않기 때문에 각각의 모듈에서 따로 관리한다.
 5. 모든 모듈은 기본 패키지로 com.aireview.review를 사용한다. 모든 클래스 및 패키지는 기본패키지 하위에 생성한다. @SpringBootApplication 어노테이션이 적용된 클래스만 기본 패키지 안에 위치하도록 해서 모든 클래스가 @ComponentScan의 대상이 되도록 한다.

## 실행 환경 구분
1. 실행 환경은 local, dev, prod  3가지로 구분된다.
2. 모듈별로 필요한 설정을 applicaion-xxx.yaml 파일에 추가한 후 spring.profiles.include로 활성화한다.
- application-db.yaml: local에서는 docker mysql 컨테이너를 실행해서 사용하고 dev, prod는 각각의 mysql db 서버를 사용한다.(추후 네이버 클라우드 서버 생성 후 application.yaml의 datasource 수정 예정)
- application-secret.yaml: jwt토큰 및 passwordencoder 설정을 위한 보안 정보를 담는다. github에 올리지 않는다.
4. local 환경에서는 서버 시작 시간을 줄이기 위해 커넥션 풀 및 스레드 풀을 최소로 유지한다.
5. 테스트 시에도 local 프로파일을 사용하도록 한다.

## 인증
- 로그인 후 발급받은 jwt로 인증 API 사용
- 로그인 시에는 UsernamePasswordAuthenticationToken으로 인증하는 DaoAuthenticationProvider로 인증
- jwt로 인증시에는 JwtAuthenticationToken으로 인증하는 JwtAuthenticationProvider로 인증
- AuthenticationManager에 DaoAuthenticationProvider와 JwtAuthenticationProvider 등록
- 필터 체인에서 인증 실패 시에는 AuthenticaitonEntryPoint에서 401 Unauthorized 반환. 인가 실패시에는 AcessDeniedHandler에서 403 Forbidden 반환

## 테스트
- domain모듈에 기본적인 필드만 포함한 user entity 및 repository를 생성하고 데이터베이스 연동 테스트
- app-api모듈 AccountControllerTest:  유저 로그인 기능 테스트

## README 작성
- 프로젝트 실행방법과 API 설계 규칙 추가.

i960107 committed:
제외된 파일
- gradle build output 관련 파일
- intellij 관련
- mac os 관련
-  f-lab에서 생성한 .gradle 폴더
- docker volumne
- 보안 정보 담은 application-secret.yaml 파일

i960107 committed:
build: init gradle

gradle 관련 파일 추가
- gradlew, gradlew.bat: macOS, windows gradle wrapper 실행 script
- gradle/: gradle 버전 및 다운로드 url 포함 property 및 gradle-wrapper 실행 파일

i960107 committed:
build: gradle 설정 추가

- root 프로젝트는 dependency 없음 -> 각 모듈별 dependency 추가
- 프로젝트 build.gradle 추가된 의존성 없지만 gradle이 프로젝트 빌드하기 위해 필요.(없으면 빌드 안됨)
- settings.gradle추가 : 포함된 하위 모듈 app-api, app-batch, domain, web.auth
  - app-api : api 어플리케이션 실행 모듈. domain 모듈을 사용. controller, service 레이어 포함
  - app-batch: batch 실행 모듈. domain 모듈을 사용.
  - domain: entity, repository 포함.
  - web: web 하위 모듈은 app에 추가되어 web관련 기능(auth, monitoring, logging)을 제공.
  - web.auth: spring security를 활용한 인증 기능 제공
 [참고] 실행가능한 모듈은 app을 붙여 구분하였고, app이 붙지 않은 모듈은 자체 실행 불가능.

i960107 committed:
build: domain 모듈 의존성 추가

1. 추가된 의존성
- spring-data-jpa
- flyway
- lombok
- mysql driver
- test
2. 실행 모듈 아니기때문에 bootJar 대신 jar task enable

i960107 committed:
docs: domain 모듈 설정추가

- profile 분리: local, dev, prod
- 추가된 설정: datasource, jpa,, flyway설정
- datasource: local(테스트 포함)에서는 mysql 사용. dev,prod는 각 환경에 따른 datasource 추가 예정
- jpa: ddl 생성은 모두 비활성화. jpa에 의한 db 스키마 변경 차단. local에서만 sql log 출력. osiv 활성화해서 뷰에서 lazy loding 가능하도록
- flyway: local, dev는 활성화. prod는 비활성화. 실제 서버에서 flyway로 직접 db 바꾸는 것  위험할 수도 있기 때문.

i960107 committed:
feat: 유저 도메인 생성

- 모듈 간 연동 검사하기 위한 유저 도메인 생성.

i960107 committed:
docs: domain 모듈 test 설정 추가

- test환경은 실행 환경과 분리되어야함. test만의 application.yaml을 가짐.
- h2 대신 local mysql 사용 : 테스트는 모두 롤백되기 때문에 상관없음. 실제 환경인 mysql와 최대한 동일하게 테스트하기 위하여 mysql 사용.

i960107 committed:
docs: mysql 실행을 위한 docker-compose.yaml 추가

- 로컬에서 테스트 시 사용되는 mysql 실행
- 프로젝트 루트에서 docker-compose.up 필요
- docker 컨테이너 생성시 프로젝트 루트에 새로운 volume(mysql/) 생성 및 데이터베이스(ai_review) 유저(review) 생성 -> application.properties에 의해 연결 가능.

i960107 committed:
test: datasource 연결 테스트

- spring context load 위한 SpringDomainTestApplication.java 추가: com.aireview.domain 패키지 하위 클래스들 component scan 대상이 됨
- entity 및 repository만 로드하기 위해 dataJpatest사용
- h2 db 대신 로컬 mysql연결을 사용하기 위해 AutoConfigureTestDatabase 어노테이션 속성 변경(datasource를 embeded source로 대체하지 않는다)

i960107 committed:
rename: rename package names

전체 모듈에 대해서 기본 패키지 com.aireview.review로 통일
모든 패키지 기본 패키지 하위에 위치.
@ComponentScan으로 별도의 설정없이 모든 모듈의 bean scan가능

i960107 committed:
rename: 웹 모듈 이름 수정

- 로깅 및 센트리 : web.monitoring
- 인증 및 인가: web.auth

i960107 committed:
rename: web 모듈 기본 구조 생성

i960107 committed:
build: jpa 전이 의존성 수정

- spring-data-jpa 모듈 의존성을 implementation -> api로 수정
- app-api의 service 레벨에서 repository 사용 가능하도록
- gradle 버전 문제로 api 사용 위해  java-library 추가

i960107 committed:
rename: rename package names

전체 모듈에 대해서 기본 패키지 com.aireview.review로 통일
모든 패키지 기본 패키지 하위에 위치.
@ComponentScan으로 별도의 설정없이 모든 모듈의 bean scan가능

i960107 committed:
chore: 테스트 설정파일 변경

main의 profile 설정 파일 바라보도록 수정.

i960107 committed:
build: app-api 모듈 추가 및 의존성 설정

- domain 모듈 추가. 추후 web 모듈 추가 예정
- 다른 모듈과 달리 app-api모듈만 Kotlin 사용: kotlin의 장점인 null check, data class를 가장 잘 활용할 수 있는 계층이라고 판단. 일부만 도입.
-추가된 의존성
1. actuator: 개발 과정에서 env, url, bean 등 확인하기 위해
2. web, validation
3. kotlin 및 kotest
4. devtools: thymeleaf 개발시 hot reload 사용위해
5. thymeleaf: view 개발용

i960107 committed:
docs: domain 모듈 설정 파일

- local, dev, prod 프로파일로 분리
- db 설정 include해서 활성화
- logging, devtool, actuator 관련 설정 추가
- dev, prod는 추후 설정 추가 예정

i960107 committed:
build: app-batch 모듈 구조 및 의존성 추가

i960107 committed:
feat: 유저 생성 및 조회 기능 추가

- 모듈 연동 확인 위한 기본 api 생성

i960107 committed:
test: 유저 생성 및 조회 기능 테스트

- random port로 실제 서버 실행 후 testRestTemplate 활용한 테스트
- 실제 서버 띄우기 때문에 transactional 적용해도 rollback 안되는 문제 있음 -> 수정 필요

i960107 committed:
refactor: 웹 모듈 삭제 및 불필요한 코드 제거

- app-api, app-batch 에 공통 적용할 web 기능들이 많지 않기 때문에 모듈 분리 효과가 떨어진다. 오히려 코드가 복잡해질 수 있다.
- flyway migraion 파일 잘못된 파일 이름 convention에 맞도록 수정
- domain 프로젝트의 test application.yaml 삭제. test시에도 local 프로파일을 사용하도록 한다. active되는 프로파일은 test 실행시에 외부에서 지정하도록 한다. dev 환경에서 test해야할 필요도 있음

i960107 committed:
refactor: docker mysql 이미지  버전 수정

- flyway 지원 안되는 5.7에서 8.0으로 버전 upgrade

i960107 committed:
build: 모듈 의존성 추가

- web 모듈 삭제에 따라 기존 Web 모듈에 있던 의존성을 app 모듈로 변경

i960107 committed:
refactor: 불필요한 @component scan 제거

모든 모듈의 base package이름을 com.aireview.review로 통일하였기 때문에 @ComponentScan 지정할 필요 없음

i960107 committed:
test: 테스트 코드 수정

- 기존 데이터 있는 mysql 데이터베이스 사용해서 테스트하기 때문에 getFirst()로 가져온 데이터 다를 수 있음
- id로 특정 데이터 가져와 검사

i960107 committed:
test: 테스트시 profile 지정

- local 프로필 사용하도록 지정. default 프로필 사용시 application-db.yaml 설정 로드 안되는 문제 발생
- random port 사용해서 테스트를 실행하고 테스트한다. @transactional은 효과가 없기 때문에 삭제한다

i960107 committed:
Fix: Spring Security 의존성 추가로 인한 문제 해결

- 기존 web.auth 모듈에 있던 Spring Security 의존성을 app-api 모듈에 추가함에 따라서 Spring Security 설정들이 활성화되었음
- 필터체인 bean을 등록하여 모든 요청을 인증 없이 사용할 수 있도록 허용하고, csrf설정을 일시적으로 비활성화한다.

i960107 committed:
fix: spring active profile을 외부에서 설정하도록 변경

i960107 committed:
docs: README 추가

- 프로젝트 실행방법
- API 설계 규칙 추가: 개발하면서 차차 업데이트 할 예정

i960107 committed:
remove: app-api 모듈 삭제

사용언어를 코틀린에서 java로 바꾸기 위해 기존 app-api 모듈 삭제

i960107 committed:
docs: .gitignore 및 docker-compose 수정

-  부하테스트 위한 docker compose  ngrinder controller / agent 추가
 - gitignore 수정
-  readme: docker nginder 컨테이너 실행 방법 추가,  인증 관련 내용 추가

i960107 committed:
docs: application-db의 local 커넥션 풀 설정 변경

- hikari cp 설정 변경: 로컬에서는 동시 요청이 많지 않기 때문에 서버 가동 시간을 줄이기 위해 connection pool 최소로 유지

i960107 committed:
docs: test 리소스 application.yaml 삭제

테스트만의 설정을 갖는게 아니라 로컬 프로필을 사용하도록 수정. (테스트 시 프로필 설정 필요)

i960107 committed:
feat: 유저 password 추가 및 빌더 제거

@빌더가 적용된 생성자를 사용하는 경우 추후에 필드가 추가될때  기존 코드에 반영 안될 위험 높음.

i960107 committed:
feat: Java 사용하는 app-api 모듈 생성

- 기존 build.gradle.kts -> build.gradle로 수정
-  application.yaml 복사

i960107 committed:
feat: 사용자 인증 기능 구현

- Spring Security의 DaoAuthenticationProvider를 상속받는 CustomDaoAuthenticaitonProvider 생성: createSuccessAuthentication() 오버라이드해서 인증 완료된 토큰의 Authentication에 AuthenticatedPrincipal(유저 아이디, 이메일 포함) 넣어주기
- jwt: 토큰 생성 및 검증
- DaoAuthenticationProvider를 사용해서 인증하기 위해서 UserDetails, UserDetailsService 구현.
- UserDetails의 authority 넣기 위한 Role enum 클래스 생성.

i960107 committed:
feat: jwt 토큰 검증 필터 추가

- Spring Security 필터 체인에 추가할 목적으로 GenericFilterBean을 구현
- 헤더에 해당 값이 없는 경우 다음 필터 체인으로 넘김
- 헤더에서 토큰을 추출하여 JwtAuthenticationToken을 만들어서 authenticationManager에게 인증을 요청
- JwtAuthneticationProvider가 JwtAuthenticationToken에 대해서 인증
- 인증을 위한 JwtAuthenticationToken에는 추출한 토큰이 credential에 들어가고, 인증 결과를 담은 JwtAuthenticationToken에는 토큰 payload 에 담긴 정보를 활용한AuthenticatedPrincipa이 principal에 추가되고 Authorities가 추가됨

i960107 committed:
feat: 필터체인에서 인증 및 인가 실패 처리 핸들러 추가

- EntryPointUnauthorizedHandler에서 인증 및 인가 실패시 http respones status 변경할 수 있도록.
- Authentication 실패시 401, Authorization 실패시 403
- 필터 체인에서 발생한 예외는 Controller Advice의 Exception Handler에 의해서 처리되지 못함.

i960107 committed:
feat: SecurityConfig 추가

- csrf 일시 비활성화: post 요청 테스트하기 위해서 추후 변경 필요
- api 별 인증 설정
- jwt사용하므로 session 생성 disable. 기본으로 매 요청마다 session 생성함.
- jwtAuthenticationFilter를 인증 시작 맨 앞(ExceptionTranslationFilter 바로 뒤)에 추가. ExceptionTranslationFilter가 인증 시작점.
- exception handling 추가
- authentication manager에 authentication provider로 CustomDaoAuthenticationProvider, JwtAuthenticationProvider 추가
  - CustomDaoAuthenticationProvider는 AccountController에서 username, password 사용해서 로그인시 사용
  - JwtAuthenticationProvider는 필터체인에서 jwt 토큰으로 인증시 사용.
- PasswordEncoder 추가

i960107 committed:
feat: 가입 및 로그인 기능 추가

i960107 committed:
feat: GlobalExceptionHandler 추가

- 로그인 과정에서 발생하는 AuthenticationException은 메시지 바디 없이 상태코드만 응답.
- 다른 예외의 경우 ErrorResponse(메시지 혹은 필드 에러)를 바디에 담아 응답

i960107 committed:
test: 로그인 테스트 추가

green-ribbon-admin commented on app-api/src/main/kotlin/com/aireview/review/controller/UserController.kt:
```diff
@@ -0,0 +1,30 @@
+        return userService.get(userId)
+    }
+
+}
```
전반적으로 nweline을 추가 해주시면 좋을거같아요~!

i960107 replied:
POSIX를 준수하기 위해서 파일 끝에 newline을 추가해 주어야하는 것을 알게 되었습니다. 다음 PR부터 newline 추가하겠습니다. 😀
그리고 README에도 코딩컨벤션으로 intellij 설정으로 자동으로 newline이 추가되도록 설정하는 법을 추가해두었습니다.

green-ribbon-admin commented on app-api/src/main/kotlin/com/aireview/review/service/UserService.kt:
```diff
@@ -0,0 +1,22 @@
+
+@Service
+class UserService(
+    @Autowired val userRepository: UserRepository
```
빈주입 시 가장 좋은방법으로 찾아봐주시면 좋을거같아요

i960107 replied:
di 받을때 테스트시 의존성 교체가 쉽도록 하고, 어플리케이션 실행시에 빈 순환 참조를 체크하려면 생성자 주입 방식이 좋다는 걸 알고는 있었는데 처음 코틀린 코드를 작성하면서 적용하지 못했습니다. 자바로 바꾸면서 생성자 주입 방식을 사용하도록 하겠습니다😀

green-ribbon-admin commented on app-api/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,38 @@
+        type:
+          descriptor:
+            sql: trace
+management:
```
액츄에이터 적용하신건가요??

green-ribbon-admin replied:
액츄에이터 설정 더 보시고 세분하게 설정 하시면 좋을거같아요

i960107 replied:
actuator 설정은 화이트리스트 방식으로 수정하고 "health" 만 사용가능하도록 허용하였습니다. 

- 기존의 web을 통해서 모든 actuator endpoing에 접근이 가능하였음
- 보안 문제 발생 가능
- 모든 endpoint들을 disable하고 사용할 endpoint만 활성화하는 white-list 방식으로 수정
- 로드밸런서의 health check 용도로 health endpoint만 활성화함.(applicaiton, datasource healthcheck 가능. diskspace는 노출되지 않도록함)
- 8081번 포트 및 base-path를 "/manage"로 수정하여 악성 공격자가 쉽게 접근하지 못하도록 수정.
- /manage/health는 헬스체크용도로 로드밸런서에 의해서 사용가능해야하므로 spring security에서 permitAll()로 인증 없이 사용가능하도록 수정

green-ribbon-admin commented on app-api/src/main/kotlin/com/aireview/review/dto/CreateUserReqDto.kt:
```diff
@@ -0,0 +1,17 @@
+    val email: String,
+    val nickname: String
+) {
+    fun toEntity(): User {
```
빌더 패턴의 문제점 때문에 다른 방식이 좋을거같아요

i960107 replied:
빌더패턴을 사용하면 추후에 필드가 추가되었을때 기존 빌더패턴을 사용해서 객체를 생성하는 코드에서 컴파일 오류가 발생하지 않아서 값이 누락될 수 있다는 것을 알게되었습니다. 빌더 패턴은 최대한 사용을 지양하겠습니다😀

green-ribbon-admin commented on docker-compose.yml:
```diff
@@ -0,0 +1,20 @@
+version: '3'
+
+services:
```
docker 인텔과 arm 기준으로 platfrom 이 달라지는게 있는데 그거 설정을 명시적으로 해주면 좋을거같아요

green-ribbon-admin commented on domain/src/main/java/com/aireview/review/domain/user/User.java:
```diff
@@ -0,0 +1,33 @@
+@Data
+public class User {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
IDENTITY vs  AUTO 의 차이점 설명 한번 해주시면 좋을거같아요

green-ribbon-admin replied:
영속성 관점에서 이점중 하나가 

동일성을 보장하는건데. 동등성과 동일성에 대해서 설명 해주세요

green-ribbon-admin replied:
spring audit aware 적용한번 부탁드려요

green-ribbon-admin replied:
p4. soft deleted 도 한번 봐주시면 좋을거같아요

green-ribbon-admin replied:
p1. spring audit aware 적용 해주세요

green-ribbon-admin commented on domain/src/main/resources/application-db.yaml:
```diff
@@ -0,0 +1,51 @@
+  config:
+    activate:
+      on-profile: local
+  datasource:
```
p1. hikaricp 에 설정에 대해 적용 부탁드립니다.

green-ribbon-admin commented on domain/src/main/resources/application-db.yaml:
```diff
@@ -0,0 +1,51 @@
+    username: review
+    password: review
+  jpa:
+    generate-ddl: false
```
p3. generate-ddl true 로 하고 ddl-auto 를 validate

i960107 replied:
flyway를 사용하면 db schema 변경이 반영되니깐 jpa에서  db schema를 변경하지 않도록 generate-ddl을 false로 설정하였습니다. 하지만 db schema와 entity가 달라도 어플리케이션이 실행되니 런타임 중에 예외가 발생할 수 있겠네요. 수정겠습니다😀

green-ribbon-admin commented on domain/src/main/resources/application-db.yaml:
```diff
@@ -0,0 +1,51 @@
+      hibernate:
+        show_sql: true
+        format_sql: true
+    open-in-view: true
```
p1. open-in-view false 로 셋팅 해주시고 영속성 컨택스트 관점에서 true, false의 차이점을 설명해주세요

i960107 replied:
영속성 컨텍스트의 엔티티 매니저를 트랜잭션을 넘어서 뷰까지 살아있도록 해서 뷰에서 원하는 객체를 지연 로딩할 수 있도록 open-in-view를 true를 설정했습니다. 컨트롤러에서 뷰 변경따라 모델을 변경하지 않아도 되도록요.😀 일단 open-in-view는 false로 막아두겠습니다.

green-ribbon-admin commented on domain/src/main/resources/db/migration/v202404181520__init.sql:
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE `user` (
+                        `id` bigint NOT NULL AUTO_INCREMENT,
```
p3. comment 를 작성해주세요

i960107 committed:
a

i960107 committed:
feat: spring actuator 설정 수정

- 기존의 web을 통해서 모든 actuator endpoing에 접근이 가능하였음
- 보안 문제 발생 가능
- 모든 endpoint들을 disable하고 사용할 endpoint만 활성화하는 white-list 방식으로 수정
- 로드밸런서의 health check 용도로 health endpoint만 활성화함.(applicaiton, datasource healthcheck 가능. diskspace는 노출되지 않도록함)
- 8081번 포트 및 base-path를 "/manage"로 수정하여 악성 공격자가 쉽게 접근하지 못하도록 수정.
- /manage/health는 헬스체크용도로 로드밸런서에 의해서 사용가능해야하므로 spring security에서 permitAll()로 인증 없이 사용가능하도록 수정

i960107 committed:
docs: 코딩컨벤션 newline추가

new line 추가를 코딩컨벤션으로 추가. 다른 사람이 작업하여도 빠뜨리지 않도록 문서화
